### ISSUE TITLE
Support LDAP Property Mappings also for LDAP Provider in Outpost

### ISSUE THREAD

**Issue by BobSilent** [goauthentik/authentik#2573]
Maybe I just did find it and this is more a question:

I wanted to extend the LDAP Provider in certain aspect
1. add an `objectClass: posixAccount` to returend users
2. change the return value of `uid` to be the User's username
3. add a new attribute `gecos` with the value of the User's name
4. add an 'calculated' attribute `homeDirectory`, which should be a combined string (base path) with the User's username

according to the [description on the terminology page](https://goauthentik.io/docs/terminology#property-mappings)

> ### Property Mappings
> Property Mappings allow you to make information available for external applications. For example, if you want to login to AWS with  authentik, you'd use Property Mappings to set the user's roles in AWS based on their group memberships in authentik.

I thought I could do it, like I already did it e.g. for the SAML Provider and "SAML Property Mapping"
but for the LDAP Provider I cannot select any of these mappings.

The description in the [Property Mappings](https://goauthentik.io/docs/property-mappings/#ldap-property-mapping)
> ### LDAP Property Mapping
> LDAP Property Mappings are used when you define a LDAP Source. 
sounds different.
states it differently and mentions the **LDAP Source** as the (only?) possibility to configure these Property mappings, so only when using and LDAP as Source not for LDAP as Provider in an Outpost.

At least from the "Property Mappings" Page I could not see that there is a difference between these mappings:
<img src="https://user-images.githubusercontent.com/13506173/159682383-ae749574-ea95-48aa-858a-450a1eadf6b6.png" width="400"/>

**Describe the solution you'd like**
Like for SAML (or OAuth Scope Mappings? did not use it right now) 
I would like to be able to map custom configurations to LDAP Attributes or extend the LDAP functionality in order to provide this information to application.

**Describe alternatives you've considered**
With some tricks I could adapt the LDAP client side with `map_attribute` in `nss_ldap.conf` but this is not always allowed, possible, well documented for client Apps.
and it currently does not help me with 4. where I wanted to add a "calculated" value

**Comment by BeryJu** [comment]
Good point, it's not signaled very clearly which kinds of property mappings work in which direction.

For the actual feature, I thought about doing this in the past but chose not to, mainly since the LDAP outpost is already quite slow, adding more computation into the API calls it does is probably not a good idea.

I would still like to implement this at some point, when the go migration is happening/has happened (when I finally have enough energy to actually start that)

**Comment by Natanji** [comment]
I want to second that giving us at least very simple options to change the output of a few of these values would be very useful. However, I see complications.

For instance, the uid is most often simply a user's login name in the LDAP world and the assumption is that it's unique and unchangable. In authentik, that's completely different - all users can change their user name on their profile page (and I think there is no way for admins to disable that?). Imagine you use LDAP auth to log into an external service, it will probably use the uid attribute. When the user changes their name, they will no longer be able to login to tat service (e.g. Nextcloud would create a new user with that uid as its internal user id).

However, using the internal UUID of authentik is *also* problematic because many LDAP enabled apps assume it to be the actual login name. For instance, this is my user page in Jellyfin after logging into it via LDAP: 
![grafik](https://user-images.githubusercontent.com/887286/160306569-034fc9ba-535c-47f3-b55e-033cc36221a8.png)
Not really a helpful username. 🤣 

It would thus be helpful if we could configure which attribute gets output to the uid field, but also, allow us to set that input field as immutable in the authentik admin panel to prevent our users from breaking things with it. Users can still change their display name.

**Comment by BobSilent** [comment]
@Natanji i know myself it is not the best, and what you are writting is really annoying also for me. And this reuqest here would really bring flexibility and the possibility to centrally preconfigure this in authentik instead of doing this in each and every application. 
> For instance, the uid is most often simply a user's login name in the LDAP world
this is also for me the biggest pain (i would also say the default setting in authentik here is not the best choice)

But if this [screenshot](https://user-images.githubusercontent.com/20963053/89155166-a73ae480-d5ab-11ea-9d28-4a9b78244053.png) is still correct, you can try changing
![image](https://user-images.githubusercontent.com/13506173/160370215-2c2ac04a-2bd5-4cba-ab16-8991d0cce6e8.png)
the LDAP Name Attribute from `uid` to `cn`.

I would assume this would already solve your issue with Jellyfin (which is also on my todo list 😉 ).

**Comment by BeryJu** [comment]
The statement about the speed is still valid, we map user attributes to LDAP attributes (as is stated on the website), however that could use some more documentation as it's currently described in a single sentence.

**Comment by gmoss-eschercloud** [comment]
I agree we have a few applications that don't allow you to change what attributes to use for the username and we are force to have really long uuid as the username's sometimes not even allowing to login due to that. 

if we could have the mapping's supported in the provider it would allow for those small changes.

**Comment by Esheads** [comment]
+1

**Comment by dminuoso** [comment]
Given that the Outpost already offers full caching for queries, do you still think performance is relevant concern here? @BeryJu 

Even if we consider non-caching, we could put up a big warning banner on a LDAP property mapping view about the adverse performance effects (in the absence of caching). That way users can decide for themselves if they are willing to trade performance for flexibility.

**Comment by derDeno** [comment]
+1

**Comment by dewi-tik** [comment]
We're not going to be implementing this due to performance concerns. But might revisit in future with a different approach which allows property mapping to be computed before hand to avoid the performance concerns.

**Comment by ericwong3** [comment]
Second this request as well. While integrating vSphere with Authentik via OpenLDAP for SSO and MFA, we came to realize that this wouldn't work because Authentik's LDAP attributes do not match the [requirement of vSphere](https://knowledge.broadcom.com/external/article?legacyId=2064977) despite having the required attributes under a slightly different name.
